lib/cn/cn_tree.c

```diff
@@ -1975,67 +1988,91 @@ cn_subspill_apply(struct subspill *ss)
 static merr_t
 cn_comp_spill(struct cn_compaction_work *w)
 {
+    uint8_t ekey[HSE_KVS_KEY_LEN_MAX];
     struct subspill *ss_saved = NULL;
     struct subspill *ss = NULL;
     struct spillctx *sctx;
     struct cn_tree *tree = w->cw_tree;
-    struct route_node *rtn = 0;
     atomic_uint *spillingp = NULL;
+    uint eklen = 0;
     merr_t err;
 
     err = cn_spill_create(w, &sctx);
     if (err)
         return err;
 
+    /* The route node pointer (rtn) is only valid while we hold the
+     * tree lock, and so we must reacquire it on each loop iteration
+     * via the edge key of the previous iteration.
+     *
+     * If rtn's edge key (rbuf) exactly matches the edge key from the
+     * last subspill (ekey) then we want the next route map node.
+     * Otherwise, the tree node from the last subspill must have been
+     * joined to the right, and so we want the route map node acquired
+     * via the edge key from the previous iteration.
+     */
     while (1) {
-        struct cn_tree_node *node;
-        struct route_node *rtnext;
-        void *lock;
         struct kvset_list_entry *le;
+        struct cn_tree_node *node;
+        struct route_node *rtn;
         uint64_t node_dgen;
-        unsigned char ekey[HSE_KVS_KEY_LEN_MAX];
-        uint eklen;
+        char *wmesg;
+        void *lock;
 
         rmlock_rlock(&tree->ct_lock, &lock);
-        rtnext = rtn ? route_node_next(rtn) : route_map_first_node(tree->ct_route_map);
-        if (!rtnext) {
+        if (eklen > 0) {
+            rtn = route_map_lookup(tree->ct_route_map, ekey, eklen);
+            if (rtn) {
+                uint8_t rbuf[HSE_KVS_KEY_LEN_MAX];
+                uint rlen;
+
+                route_node_keycpy(rtn, rbuf, sizeof(rbuf), &rlen);
+                if (rlen == eklen && 0 == memcmp(rbuf, ekey, eklen))
+                    rtn = route_node_next(rtn);
+                else
+                    ev_warn(1);
+            }
+        } else {
+            rtn = route_map_first_node(tree->ct_route_map);
+        }
```

Comment:
I'm going to remove that ev_warn()..  @gsramdasi please scrutinize this route-next logic, both @nabeelmmd and I have reviewed it and agree that it's correct.